Userland/Applications/HexEditor/HexEditorWidget.cpp

```diff
@@ -359,52 +359,46 @@ void HexEditorWidget::update_inspector_values(size_t position)
         value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::Double, "");
     }
 
-    // FIXME: This probably doesn't honour endianness correctly.
     Utf8View utf8_view { ReadonlyBytes { reinterpret_cast<u8 const*>(&unsigned_64_bit_int), 4 } };
     size_t valid_bytes;
     utf8_view.validate(valid_bytes);
     if (valid_bytes == 0)
-        value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF8, "");
+        value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF8, make_display_string(""));
     else
-        value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF8, utf8_view.unicode_substring_view(0, 1).as_string());
+        value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF8, make_display_string(utf8_view.unicode_substring_view(0, 1).as_string()));
 
+    // FIXME: The byte read count doesn't need to be even for UTF-16 to be valid
     if (byte_read_count % 2 == 0) {
         Utf16View utf16_view { Span<u16 const> { reinterpret_cast<u16 const*>(&unsigned_64_bit_int), 4 } };
         size_t valid_code_units;
         utf16_view.validate(valid_code_units);
         if (valid_code_units == 0)
-            value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF16, "");
+            value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF16, make_display_string(""));
         else
-            value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF16, utf16_view.unicode_substring_view(0, 1).to_deprecated_string().release_value_but_fixme_should_propagate_errors());
+            value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF16, make_display_string(utf16_view.unicode_substring_view(0, 1).to_deprecated_string().release_value_but_fixme_should_propagate_errors()));
     } else {
         value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::UTF16, "");
     }
 
     StringBuilder ascii_string_builder;
     for (size_t i = 0; i < selected_bytes.size(); i++) {
         char byte_char = static_cast<char>(selected_bytes[i]);
-
-        if (!is_ascii_control(byte_char) && is_ascii(byte_char)) {
-            ascii_string_builder.append(byte_char);
-        } else {
-            ascii_string_builder.append(' ');
-        }
+        ascii_string_builder.append(byte_char);
```

Comment:
This will append valid UTF-8, which we don't want since it's supposed to be an ASCII string. The control character replacement will not fix this.

Reply:
Hmm I think we're good here, because we're going byte by byte

At the single byte level, ASCII == UTF-8. If we append a byte that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior here

Let me know if you had something else in mind

Reply:
Hmm I think we're good here, because we're going byte by byte

At the single byte level, ASCII == UTF-8. If we append a byte that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior

Let me know if you had something else in mind

Reply:
> At the single byte level, ASCII == UTF-8. If we append a byte (i.e. char) that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior
>
> I think we would only run into the issue you lay out here if we were appending a _code point_

The problem, as you correctly identify, is with non-ASCII multibyte sequences. In UTF-8, many of those are valid single (!) codepoints, while in ASCII, they are many invalid characters. This is an important distinction to make if the format the user is working with is explicitly requiring valid ASCII and e.g. the user tries to figure out why some string is not accepted by a reader program.

I think it is very important to be strict here. A field that says "ASCII" should only parse and display valid ASCII as defined by ANSI and ISO many decades ago. Because many (especially old) parsers will be strict about this and even if we are strict, the user can still decide to ignore invalid ASCII characters if need be.

Reply:
Sorry to say, but I _really_ don't grok what you are saying here

If you could give an example, that would go a really long way to help me understand the point you are getting at

Reply:
The claim I believe you are making is that there can be a byte that is not valid ASCII that will be shown in the ASCII display view with the current logic

If that is correct, please give me an example of such a byte that will produce the bug. If that's not correct, please do clarify what you mean